File: MorySnip/Classes/Editor.vb
Imports System.Drawing.Drawing2D

Public Class Editor
    Dim PreviousSteps As New List(Of Image)
    Dim NextSteps As New List(Of Image)
    Dim Layers As New List(Of Layer)
    Dim ImagePosition As New Point(0, 0)

    Dim LastButton As MouseButtons = MouseButtons.None
    Dim _LastNumber As Integer = 1

    Public Event LastNumberChanged(sender As Object, e As EventArgs)

    Public Sub CreateCheckpoint(Optional Image As Image = Nothing)
        If Image Is Nothing Then
            Image = Me.GetResult()
        End If
        Me.PreviousSteps.Add(Image)
        Me.NextSteps.Clear()
    End Sub

    Public Property LastNumber As Integer
        Get
            Return Me._LastNumber
        End Get
        Set(value As Integer)
            Me._LastNumber = value

            RaiseEvent LastNumberChanged(Me, New EventArgs())
        End Set
    End Property

    Public ReadOnly Property CanRedo As Boolean
        Get
            Return Me.NextSteps.Count > 0
        End Get
    End Property

    Public ReadOnly Property CanUndo As Boolean
        Get
            Return Me.PreviousSteps.Count > 0
        End Get
    End Property

    Public Sub Redo()
        If Me.CanRedo Then
            Me.PreviousSteps.Add(Me.GetResult())
            Me.BackgroundImage = Me.NextSteps(Me.NextSteps.Count - 1)
            Me.NextSteps.RemoveAt(Me.NextSteps.Count - 1)
        End If
    End Sub

    Public Sub Undo()
        If Me.CanUndo Then
            Me.NextSteps.Add(Me.GetResult())
            Me.Layers.Clear()
            Me.BackgroundImage = Me.PreviousSteps(Me.PreviousSteps.Count - 1)
            Me.PreviousSteps.RemoveAt(Me.PreviousSteps.Count - 1)
        End If
    End Sub

    Public Sub Render()
        Dim result As Image = Me.GetResult()

        Me.BackgroundImage = result

        Me.Layers.Clear()
        Me.ImagePosition = New Point()
    End Sub

    Public Function GetResult() As Image
        Dim result As New Bitmap(Me.Width, Me.Height)

        Dim g As Graphics = Graphics.FromImage(result)

        g.Clear(Color.White)

        If Me.Img IsNot Nothing Then
            g.DrawImage(Me.Img, Me.ImagePosition.X, Me.ImagePosition.Y, Me.Img.Size.Width, Me.Img.Size.Height)
        End If

        For Each l As Layer In Me.Layers
            l.Render(g)
        Next
        Return result
    End Function
    Public ReadOnly Property CurrentPen As New Pen(Brushes.Red, 2) With {.StartCap = LineCap.Round, .EndCap = LineCap.Round, .Alignment = PenAlignment.Inset}

    Dim CurrentBrushValue As New SolidBrush(Color.Red)
    Public Property CurrentBrush As Brush
        Get
            Return Me.CurrentBrushValue
        End Get
        Set(value As Brush)
            Me.CurrentBrushValue = value
        End Set
    End Property

    Private Property Img() As Bitmap
        Get
            Return Me.BackgroundImage
        End Get
        Set(ByVal value As Bitmap)
            Me.ImagePosition.X = 0
            Me.ImagePosition.Y = 0
            Me.BackgroundImage = value
        End Set
    End Property

    Dim GrayedAreaBrush As New SolidBrush(Color.FromArgb(200, 0, 0, 0))
    Dim HighlightColor As Color = Color.FromArgb(100, Color.Yellow)
    Dim HighlightBrush As New SolidBrush(Me.HighlightColor)

    Protected Overrides Sub OnPaintBackground(ByVal e As PaintEventArgs)
        Dim g As Graphics = e.Graphics

        g.Clear(Color.White)

        If Me.Img Is Nothing Then
            g.DrawString("NO IMAGE !!!", Me.Font, Brushes.Red, 5, 5)
        Else
            g.DrawImage(Me.Img, Me.ImagePosition.X, Me.ImagePosition.Y, Me.Img.Size.Width, Me.Img.Size.Height)
        End If

        For Each l As Layer In Me.Layers
            l.Render(g)
        Next

        If Me.NewLayer IsNot Nothing Then
            Me.NewLayer.Paint(g)
        End If
    End Sub

    Public Enum PaintModes
        Free
        Line
        Arrow
        Oval
        Rect
        Number

        Highlight
        Invert
        Blur
        Puzzle
        Crop
        Grayscale
    End Enum

    Dim _PaintMode As PaintModes = PaintModes.Free
    Dim _PaintModeLast As PaintModes = PaintModes.Free

    Public Sub SetLastPaintMode()
        Me.PaintMode = Me._PaintModeLast
    End Sub

    Public Property PaintMode() As PaintModes
        Get
            Return Me._PaintMode
        End Get
        Set(ByVal value As PaintModes)
            Me._PaintModeLast = Me._PaintMode
            Me._PaintMode = value
        End Set
    End Property

    Public Property FillObjecs() As Boolean = True

    Dim Painting As Boolean = False
    Dim PBegin As Point
    Dim PEnd As Point

    Private Sub Editor_BackgroundImageChanged(sender As Object, e As EventArgs) Handles Me.BackgroundImageChanged
        Me.ResetImageSizeAndPosition()
    End Sub

    Public Sub ResetImageSizeAndPosition()
        If Me.BackgroundImage IsNot Nothing Then
            Me.Size = Me.BackgroundImage.Size
            Me.ImagePosition = New Point()
            Me.Refresh()
        End If
    End Sub

    Private Sub Editor_KeyDown(sender As Object, e As KeyEventArgs) Handles Me.KeyDown
        If e.KeyCode = Keys.Escape AndAlso e.Control = False Then
            Me.NewLayer = Nothing
            Me.Refresh()
        End If
    End Sub

    Dim NewLayer As Layer

    Private Sub Editor_MouseClick(sender As Object, e As MouseEventArgs) Handles Me.MouseClick
        If Me.LastButton = MouseButtons.Left AndAlso e.Button = MouseButtons.Right OrElse Me.LastButton = MouseButtons.Right AndAlso e.Button = MouseButtons.Left Then
            If TypeOf Me.NewLayer Is Layer_Arrow Then
                Dim NewLayerArrow As Layer_Arrow = Me.NewLayer
                Select Case NewLayerArrow.ArrowMode
                    Case ArrowModes.AtEnd
                        NewLayerArrow.ArrowMode = ArrowModes.AtStart
                    Case ArrowModes.AtStart
                        NewLayerArrow.ArrowMode = ArrowModes.Both
                    Case ArrowModes.Both
                        NewLayerArrow.ArrowMode = ArrowModes.AtEnd
                End Select
            End If
        End If
    End Sub

    Private Sub Editor_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs) Handles Me.MouseDown
        If Me.LastButton = MouseButtons.None Then
            Me.LastButton = e.Button

            Select Case e.Button
                Case MouseButtons.Left, MouseButtons.Right
                    Select Case Me.PaintMode
                        Case PaintModes.Line
                            If e.Button = MouseButtons.Right Then
                                Me.CurrentPen.DashStyle = DashStyle.DashDotDot
                            End If
                            Me.NewLayer = New Layer_Line(Me.CurrentPen, e.Location)
                        Case PaintModes.Rect
                            Me.NewLayer = New Layer_Rect(Me.CurrentPen, Me.CurrentBrush, e.Location) With {.Fill = e.Button = MouseButtons.Right Xor Me.FillObjecs}
                        Case PaintModes.Number
                            Me.NewLayer = New Layer_Number(Me.CurrentPen, Me.CurrentBrush, e.Location, Me.LastNumber) With {.Fill = e.Button = MouseButtons.Right Xor Me.FillObjecs}
                        Case PaintModes.Oval
                            Me.NewLayer = New Layer_Oval(Me.CurrentPen, Me.CurrentBrush, e.Location) With {.Fill = e.Button = MouseButtons.Right Xor Me.FillObjecs}
                        Case PaintModes.Free
                            Me.NewLayer = New Layer_Free(Me.CurrentPen, Me.CurrentBrush, e.Location) With {.Fill = e.Button = MouseButtons.Right Xor Me.FillObjecs}
                        Case PaintModes.Arrow
                            Me.NewLayer = New Layer_Arrow(Me.CurrentPen, Me.CurrentBrush, e.Location, ArrowMode:=If(e.Button = MouseButtons.Right, ArrowModes.AtStart, ArrowModes.AtEnd))
                        Case PaintModes.Invert
                            Me.NewLayer = New Layer_Action(e.Location, Actions.Invert, IIf(e.Button = MouseButtons.Right, Zones.NotSelected, Zones.Selected))
                        Case PaintModes.Blur
                            Me.NewLayer = New Layer_Action(e.Location, Actions.Blur, IIf(e.Button = MouseButtons.Right, Zones.NotSelected, Zones.Selected))
                        Case PaintModes.Puzzle
                            Me.NewLayer = New Layer_Action(e.Location, Actions.Puzzle, IIf(e.Button = MouseButtons.Right, Zones.NotSelected, Zones.Selected))
                        Case PaintModes.Grayscale
                            Me.NewLayer = New Layer_Action(e.Location, Actions.Grayscale, IIf(e.Button = MouseButtons.Right, Zones.NotSelected, Zones.Selected))
                        Case PaintModes.Highlight
                            Me.NewLayer = New Layer_Action(e.Location, Actions.Highlight, Zones.Selected)
                        Case PaintModes.Crop
                            Me.NewLayer = New Layer_Action(e.Location, Actions.Crop, Zones.Selected)
                    End Select

                Case MouseButtons.Middle
                    Me.PBegin = e.Location
                Case Else
                    Me.LastButton = MouseButtons.None
            End Select
        End If
    End Sub

    Private Sub Editor_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs) Handles Me.MouseMove
        If e.Button = Me.LastButton And Not Me.LastButton = MouseButtons.None Then
            Select Case e.Button
                Case MouseButtons.Left, MouseButtons.Right
                    If Me.NewLayer IsNot Nothing Then
                        Me.NewLayer.Step(e.Location)
                    End If
                Case MouseButtons.Middle
                    Me.ImagePosition.X += (e.Location - Me.PBegin).X
                    Me.ImagePosition.Y += (e.Location - Me.PBegin).Y

                    Me.PBegin = e.Location
            End Select

            Me.Refresh()
        End If
    End Sub

    Private Sub Editor_MouseUp(ByVal sender As Object, ByVal e As MouseEventArgs) Handles Me.MouseUp
        If e.Button = Me.LastButton And Not Me.LastButton = MouseButtons.None Then
            Select Case e.Button
                Case MouseButtons.Left, MouseButtons.Right
                    If Me.NewLayer IsNot Nothing Then
                        Me.NewLayer.Stop(e.Location)
                        Me.CreateCheckpoint()

                        If TypeOf Me.NewLayer Is Layer_Action Then
                            Dim ActionLayer As Layer_Action = Me.NewLayer
                            Me.Render()
                            Me.BackgroundImage = ApplyAction(Me.BackgroundImage, ActionLayer.Action, ActionLayer.Zone, ActionLayer.Bounds)
                        Else
                            If TypeOf Me.NewLayer Is Layer_Number Then
                                Me.LastNumber += 1
                            End If
                            If TypeOf Me.NewLayer Is Layer_Line AndAlso Me.LastButton = MouseButtons.Right Then
                                Me.CurrentPen.DashStyle = DashStyle.Solid
                            End If
                            Me.Layers.Add(Me.NewLayer)
                        End If

                        Me.NewLayer = Nothing
                    End If
                Case MouseButtons.Middle
                    Me.ImagePosition.X += (e.Location - Me.PBegin).X
                    Me.ImagePosition.Y += (e.Location - Me.PBegin).Y

                    Me.PBegin = e.Location
            End Select

            Me.LastButton = MouseButtons.None

            Me.Refresh()
        End If
    End Sub

    Public Sub RotateFlip(ByVal value As RotateFlipType)
        Me.Render()
        Me.Img.RotateFlip(value)
        Me.ResetImageSizeAndPosition()
    End Sub

    Private Sub Editor_MouseWheel(sender As Object, e As MouseEventArgs) Handles Me.MouseWheel
        If Me.NewLayer IsNot Nothing Then
            If e.Delta > 0 Then
                If Me.CurrentPen.Width < 30 Then Me.CurrentPen.Width += 1
            Else
                If Me.CurrentPen.Width > 1 Then Me.CurrentPen.Width -= 1
            End If
            Me.Refresh()
            CType(e, HandledMouseEventArgs).Handled = True
        End If
    End Sub
End Class

File: MorySnip/Classes/Editor.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Linq;
using System.Windows.Forms;
using Morysoft.MorySnip.Classes;
using Morysoft.MorySnip.Classes.Layers;
using Morysoft.MorySnip.Draw;
using Morysoft.MorySnip.Modules;

namespace Morysoft.MorySnip;

public delegate void LastNumberChangedEventHandler(object sender, EventArgs e);

public partial class Editor
{
    private readonly Stack<EditorStep> previousSteps = new();
    private readonly Stack<EditorStep> nextSteps = new();
    private readonly List<Layer> Layers = new();
    private Point imagePosition = new(0, 0);

    private MouseButtons LastButton = MouseButtons.None;
    private int lastNumber = 1;

    private EditorPaintMode paintMode = EditorPaintMode.Arrow;
    private EditorPaintMode paintModeLast = EditorPaintMode.Arrow;

    private Point startPoint;

    private Layer? newLayer;

    public event LastNumberChangedEventHandler? LastNumberChanged;

    public int LastNumber
    {
        get => this.lastNumber;
        set
        {
            this.lastNumber = value;

            LastNumberChanged?.Invoke(this, EventArgs.Empty);
        }
    }

    public bool CanRedo => this.nextSteps.Count > 0;

    public bool CanUndo => this.previousSteps.Count > 0;

    public string QuickText { get; set; } = String.Empty;

    public EditorPaintMode PaintMode
    {
        get => this.paintMode;
        set
        {
            this.paintModeLast = this.paintMode;
            this.paintMode = value;
        }
    }

    public Editor()
    {
        this.InitializeComponent();
    }

    public void Redo()
    {
        if (this.CanRedo)
        {
            this.previousSteps.Push(this.CreateStep());

            var step = this.nextSteps.Pop();

            this.BackgroundImage = step.Image;
            this.LastNumber = step.LastNumber;
        }
    }

    public void Undo()
    {
        if (this.CanUndo)
        {
            this.nextSteps.Push(this.CreateStep());

            this.Layers.Clear();

            var step = this.previousSteps.Pop();

            this.ApplyStep(step);
        }
    }

    public EditorStep CreateStep() => new(this.GetResult(), this.LastNumber, this.Layers.Select(l => l).ToList()); //TODO: clone every layer

    public void ApplyStep(EditorStep step)
    {
        if (step is null)
        {
            throw new ArgumentNullException(nameof(step));
        }

        this.BackgroundImage = step.Image;
        this.LastNumber = step.LastNumber;
    }

    public void Render()
    {
        var result = this.GetResult();

        this.BackgroundImage = result;

        this.Layers.Clear();
        this.imagePosition = new Point();
    }

    public Bitmap GetResult()
    {
        var result = new Bitmap(this.Width, this.Height);

        var g = Graphics.FromImage(result);

        g.Clear(Color.White);

        if (this.EditableImage != null)
        {
            g.DrawImage(this.EditableImage, this.imagePosition.X, this.imagePosition.Y, this.EditableImage.Size.Width, this.EditableImage.Size.Height);
        }

        foreach (var l in this.Layers)
        {
            l.Render(g);
        }

        return result;
    }

    public Pen CurrentPen { get; private set; } = new Pen(Brushes.Red, 2) { StartCap = LineCap.Round, EndCap = LineCap.Round, Alignment = PenAlignment.Inset };

    public Brush CurrentBrush { get; set; } = new SolidBrush(Color.Red);

    private Image EditableImage
    {
        get => this.BackgroundImage;
        set
        {
            this.imagePosition.X = 0;
            this.imagePosition.Y = 0;
            this.BackgroundImage = value;
        }
    }

    public void SetLastPaintMode() => this.PaintMode = this.paintModeLast;

    public void RotateFlip(RotateFlipType value)
    {
        this.Render();
        this.EditableImage.RotateFlip(value);
        this.ResetImageSizeAndPosition();
    }

    public bool FillObjecs { get; set; } = true;

    protected override void OnPaintBackground(PaintEventArgs e)
    {
        if (e is null)
        {
            throw new ArgumentNullException(nameof(e));
        }

        var g = e.Graphics;

        g.Clear(Color.White);

        if (this.EditableImage is null)
        {
            g.DrawString("NO IMAGE !!!", this.Font, Brushes.Red, 5, 5);
        }
        else
        {
            g.DrawImage(this.EditableImage, this.imagePosition.X, this.imagePosition.Y, this.EditableImage.Size.Width, this.EditableImage.Size.Height);
        }

        foreach (var l in this.Layers)
        {
            l.Render(g);
        }

        if (this.newLayer is not null)
        {
            this.newLayer.Paint(g);
        }
    }

    private void Editor_BackgroundImageChanged(object sender, EventArgs e) => this.ResetImageSizeAndPosition();

    public void ResetImageSizeAndPosition()
    {
        if (this.BackgroundImage is not null)
        {
            this.Size = this.BackgroundImage.Size;
            this.imagePosition = new Point();
            this.Refresh();
        }
    }

    private void Editor_KeyDown(object sender, KeyEventArgs e)
    {
        if (e.Control || e.Shift || this.newLayer is null)
        {
            return;
        }

        if (e.KeyCode == Keys.Escape)
        {
            this.newLayer = null;
            this.Refresh();
        }
        else if (e.KeyCode == Keys.Oem3) // `/~ key
        {
            this.SecondaryAction();
        }
        else if (e.KeyCode == Keys.S && this.newLayer.Pen is not null)
        {
            SwitchPenDashStyle(this.newLayer.Pen);
            this.Refresh();
        }
    }

    private void CreateCheckpoint()
    {
        this.previousSteps.Push(this.CreateStep());
        this.nextSteps.Clear();
    }

    private void Editor_MouseClick(object sender, MouseEventArgs e)
    {
        if (this.LastButton == MouseButtons.Left && e.Button == MouseButtons.Right || this.LastButton == MouseButtons.Right && e.Button == MouseButtons.Left)
        {
            this.SecondaryAction();
        }
    }

    private void SecondaryAction()
    {
        switch (this.newLayer)
        {
            case LayerArrow layerArrow:
                {
                    layerArrow.ArrowMode = layerArrow.ArrowMode switch
                    {
                        ArrowModes.AtEnd => ArrowModes.AtStart,
                        ArrowModes.AtStart => ArrowModes.Both,
                        ArrowModes.Both => ArrowModes.AtEnd,
                        _ => ArrowModes.AtStart,
                    };
                    break;
                }
            case LayerAction layerAction:
                layerAction.Zone = layerAction.Zone switch
                {
                    Zones.Selected => Zones.NotSelected,
                    Zones.NotSelected => Zones.Selected,
                    _ => Zones.All,
                };
                break;
            case LayerLine layerLine:
                if (layerLine.Pen is null)
                {
                    return;
                }

                SwitchPenDashStyle(layerLine.Pen);
                break;
            case Layer layer:
                layer.Fill ^= true;
                break;
        }

        this.Refresh();
    }

    private static void SwitchPenDashStyle(Pen pen)
    {
        pen.DashStyle = pen.DashStyle switch
        {
            DashStyle.Solid => DashStyle.Dash,
            DashStyle.Dash => DashStyle.Dot,
            _ => DashStyle.Solid,
        };
    }

    private void Editor_MouseDown(object sender, MouseEventArgs e)
    {
        if (this.LastButton == MouseButtons.None)
        {
            this.LastButton = e.Button;

            switch (e.Button)
            {
                case MouseButtons.Left:
                case MouseButtons.Right:
                    {
                        this.BeginLayer(e);

                        break;
                    }

                case MouseButtons.Middle:
                    {
                        this.startPoint = e.Location;
                        break;
                    }

                default:
                    {
                        this.LastButton = MouseButtons.None;
                        break;
                    }
            }
        }
    }

    private void BeginLayer(MouseEventArgs e)
    {
        switch (this.PaintMode)
        {
            case EditorPaintMode.Line:
                {
                    if (e.Button == MouseButtons.Right)
                    {
                        this.CurrentPen.DashStyle = DashStyle.Dash;
                    }

                    this.newLayer = new LayerLine(this.CurrentPen, e.Location);
                    break;
                }

            case EditorPaintMode.Rect:
                {
                    this.newLayer = new LayerRect(this.CurrentPen, this.CurrentBrush, e.Location) { Fill = e.Button == MouseButtons.Right ^ this.FillObjecs };
                    break;
                }

            case EditorPaintMode.Number:
                {
                    this.newLayer = new LayerNumber(this.CurrentPen, this.CurrentBrush, e.Location, this.LastNumber) { Fill = e.Button == MouseButtons.Right ^ this.FillObjecs };
                    break;
                }

            case EditorPaintMode.Oval:
                {
                    this.newLayer = new LayerOval(this.CurrentPen, this.CurrentBrush, e.Location) { Fill = e.Button == MouseButtons.Right ^ this.FillObjecs };
                    break;
                }

            case EditorPaintMode.Free:
                {
                    this.newLayer = new LayerFree(this.CurrentPen, this.CurrentBrush, e.Location) { Fill = e.Button == MouseButtons.Right ^ this.FillObjecs };
                    break;
                }

            case EditorPaintMode.Arrow:
                {
                    this.newLayer = new LayerArrow(this.CurrentPen, this.CurrentBrush, e.Location, arrowMode: e.Button == MouseButtons.Right ? ArrowModes.AtStart : ArrowModes.AtEnd);
                    break;
                }

            case EditorPaintMode.Magnifier:
                {
                    this.newLayer = new LayerMagnifier(
                        this.CurrentPen,
                        this.CurrentBrush,
                        (Bitmap)this.BackgroundImage,
                        e.Location,
                        2,
                        50
                    );
                    break;
                }

            case EditorPaintMode.Text:
                {
                    this.newLayer = new LayerText(
                        this.QuickText, this.CurrentPen, this.CurrentBrush, e.Location, this.Font,
                        e.Button == MouseButtons.Right ? ArrowModes.AtStart : ArrowModes.AtEnd);
                    break;
                }

            case EditorPaintMode.Invert:
                {
                    this.newLayer = new LayerAction(e.Location, Actions.Invert, e.Button == MouseButtons.Right ? Zones.NotSelected : Zones.Selected);
                    break;
                }

            case EditorPaintMode.Blur:
                {
                    this.newLayer = new LayerAction(e.Location, Actions.Blur, e.Button == MouseButtons.Right ? Zones.NotSelected : Zones.Selected);
                    break;
                }

            case EditorPaintMode.Puzzle:
                {
                    this.newLayer = new LayerAction(e.Location, Actions.Puzzle, e.Button == MouseButtons.Right ? Zones.NotSelected : Zones.Selected);
                    break;
                }

            case EditorPaintMode.Grayscale:
                {
                    this.newLayer = new LayerAction(e.Location, Actions.Grayscale, e.Button == MouseButtons.Right ? Zones.NotSelected : Zones.Selected);
                    break;
                }

            case EditorPaintMode.Highlight:
                {
                    this.newLayer = new LayerAction(e.Location, Actions.Highlight, Zones.Selected);
                    break;
                }

            case EditorPaintMode.Crop:
                {
                    this.newLayer = new LayerAction(e.Location, Actions.Crop, Zones.Selected);
                    break;
                }
        }
    }

    private void Editor_MouseMove(object sender, MouseEventArgs e)
    {
        if (e.Button == this.LastButton & !(this.LastButton == MouseButtons.None))
        {
            switch (e.Button)
            {
                case MouseButtons.Left:
                case MouseButtons.Right:
                    {
                        if (this.newLayer != null)
                        {
                            this.newLayer.Step(e.Location);
                        }

                        break;
                    }

                case MouseButtons.Middle:
                    {
                        this.imagePosition.X += e.Location.X - this.startPoint.X;
                        this.imagePosition.Y += e.Location.Y - this.startPoint.Y;

                        this.startPoint = e.Location;

                        break;
                    }
            }

            this.Refresh();
        }
    }

    private void Editor_MouseUp(object sender, MouseEventArgs e)
    {
        if (e.Button == this.LastButton & !(this.LastButton == MouseButtons.None))
        {
            switch (e.Button)
            {
                case MouseButtons.Left:
                case MouseButtons.Right:
                    {
                        this.CompleteLayer(e);

                        break;
                    }

                case MouseButtons.Middle:
                    {
                        this.imagePosition.X += (e.Location.X - this.startPoint.X);
                        this.imagePosition.Y += (e.Location.Y - this.startPoint.Y);

                        this.startPoint = e.Location;
                        break;
                    }
            }

            this.LastButton = MouseButtons.None;

            this.Refresh();
        }
    }

    private void CompleteLayer(MouseEventArgs e)
    {
        if (this.newLayer == null)
        {
            return;
        }

        this.newLayer.Stop(e.Location);

        if (!this.newLayer.IsValid)
        {
            this.newLayer = null;
            return;
        }

        this.CreateCheckpoint();

        if (this.newLayer is LayerAction ActionLayer)
        {
            this.Render();
            this.BackgroundImage = Helpers.ApplyAction((Bitmap)this.BackgroundImage, ActionLayer.Action, ActionLayer.Zone, ActionLayer.Bounds);
        }
        else
        {
            if (this.newLayer is LayerNumber)
            {
                this.LastNumber += 1;
            }

            this.CurrentPen.DashStyle = DashStyle.Solid;

            this.Layers.Add(this.newLayer);
        }

        this.newLayer = null;
    }

    private void Editor_MouseWheel(object sender, MouseEventArgs e)
    {
        if (this.newLayer == null)
        {
            return;
        }

        if (e.Delta > 0)
        {
            if (this.CurrentPen.Width < 30)
            {
                this.CurrentPen.Width += 1;
            }
        }
        else if (this.CurrentPen.Width > 1)
        {
            this.CurrentPen.Width -= 1;
        }

        this.Refresh();

        if (e is HandledMouseEventArgs he)
        {
            he.Handled = true;
        }
    }
}
